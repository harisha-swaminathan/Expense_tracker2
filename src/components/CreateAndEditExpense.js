import React, { useState } from "react";
import DatePicker from "react-datepicker";

const CreateAndEditExpense = (props) => {
    const expenseToEdit = props.expenseToEdit || {};    
    const [description, setDescription] =  useState(expenseToEdit.description ? expenseToEdit.description : '');
    const [amount, setAmount] =  useState(expenseToEdit.amount ?  (expenseToEdit.amount/100).toString() : '');
    const [note, setNote] =  useState(expenseToEdit.note ? expenseToEdit.note : '');
    const [createdAt, setCreatedAt] =  useState(expenseToEdit.createdAt || new Date());
    const [error, setError] =  useState('');

    const onDescriptionChange = (e) => {
        setDescription(e.target.value)
    };
    const onAmountChange = (e) => {
        const amount = e.target.value;

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            setAmount(e.target.value);
        }
    };
    const onNoteChange = (e) => {
        setNote(e.target.value)
    };
    const onDateChange = (createdAt) => {
        if(createdAt) {
            setCreatedAt(createdAt);
        }
    };
    const onExpenseFormSubmit = (e) => {
        e.preventDefault();
        if(!description || !amount) {
            setError('Please provide Description and amount!');
            return;
        }
        setError('');
        props.onSubmit({
            description,
            amount: parseFloat(amount, 10) * 100,
            createdAt: createdAt.valueOf(),
            note
        });
    };
    return (
        <>
            {!!error && <p>{error}</p>}
            <form onSubmit={onExpenseFormSubmit} data-testid="form-element">
                <input
                    type="text"
                    placeholder="Description"
                    onChange={onDescriptionChange}
                    value = {description}
                    autoFocus
                />
                <input
                    type="text"
                    placeholder="Amount"
                    onChange={onAmountChange}
                    value = {amount}
                />
                <DatePicker 
                    selected={createdAt} 
                    showIcon
                    format="dd/MM/yy"
                    onChange={onDateChange} 
                />
                <textarea
                    placeholder="add a note for you expense"
                    onChange={onNoteChange}
                    value = {note}
                />
                <button data-testid="form-submit"> Add Expense</button>
            </form>
        </>
    );
}

export default CreateAndEditExpense;

