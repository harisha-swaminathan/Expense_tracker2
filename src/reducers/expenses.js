const expenseReducerDefaultState = [];

const expensesReducer = (state = expenseReducerDefaultState, action) => {
    switch(action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'DELETE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if(expense.id === action.id){
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense;
                }
            })
        default:
            return state;
    }
};

export default expensesReducer;