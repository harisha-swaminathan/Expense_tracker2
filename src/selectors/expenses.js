import { isBefore, isAfter } from "date-fns";


const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
      
        const description = expense.description.toLowerCase();
        const textMatch = typeof text !== 'string' || description.includes(text.toLowerCase());
        const startDateMatch = !startDate || isAfter(expense.createdAt, startDate); 
        const endDateMatch =  !endDate || isBefore(expense.createdAt, endDate);

        return textMatch && startDateMatch && endDateMatch;
    }).sort((expense1, expense2) => {
        if(sortBy == 'date'){
            return expense1.createdAt < expense2.createdAt ? 1 : -1;
        } else {
            return expense1.amount < expense2.amount ? 1 : -1;
        }
    })
};

export default getVisibleExpenses;