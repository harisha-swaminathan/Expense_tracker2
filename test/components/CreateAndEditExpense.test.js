import React from 'react';
import { render, screen, fireEvent} from '@testing-library/react';
import '@testing-library/jest-dom';
import { userEvent } from '@testing-library/user-event';
import CreateAndEditExpense from "../../src/components/CreateAndEditExpense";
import { BrowserRouter } from 'react-router-dom';
import expenses from '../fixtures/expenses';

let mockedDate;
let dateSpy;

beforeEach(() => {
    mockedDate = new Date(1996, 6, 19);
    dateSpy = jest.spyOn(global, "Date").mockImplementation(() => mockedDate);
});
  
afterEach(() => {
    dateSpy.mockRestore();
});

test('expenseList is rendered with expenses as expected', () =>{
    const mockedDate = new Date(1996, 6, 19);
    const { asFragment } = render(<BrowserRouter> <CreateAndEditExpense expenseToEdit={expenses[0]}/> </BrowserRouter>);

    expect(asFragment()).toMatchSnapshot();
});

test('expenseList is rendered without any expense as expected', () =>{
    const { asFragment } = render(<BrowserRouter> <CreateAndEditExpense/> </BrowserRouter>);

    expect(asFragment()).toMatchSnapshot();
});

test('should render error', async () => {
    const { asFragment } = render(<BrowserRouter> <CreateAndEditExpense/> </BrowserRouter>);
    const user = userEvent.setup();
    const submitButton = screen.getByTestId('form-submit');

    await user.click(submitButton);

    expect(screen.getByText('Please provide Description and amount!')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
});

test('should set description on input change', async () => {
    const { asFragment } = render(<BrowserRouter> <CreateAndEditExpense/> </BrowserRouter>);
    const description = screen.getByPlaceholderText('Description');

    fireEvent.change(description, {target: {value: 'new description'}});
    expect(description.value).toBe('new description');
    expect(asFragment()).toMatchSnapshot();
});

test('should set notes on input change', async () => {
    const { asFragment } = render(<BrowserRouter> <CreateAndEditExpense/> </BrowserRouter>);
    const notes = screen.getByPlaceholderText('add a note for you expense');

    fireEvent.change(notes, {target: {value: 'a new bag for my bday'}});
    expect(notes.value).toBe('a new bag for my bday');
    expect(asFragment()).toMatchSnapshot();
});

test('should set amount on valid input', async () => {
    const { asFragment } = render(<BrowserRouter> <CreateAndEditExpense/> </BrowserRouter>);
    const amount = screen.getByPlaceholderText('Amount');
    const value = '12.50';

    fireEvent.change(amount, {target: {value}});
    expect(amount.value).toBe(value);
    expect(asFragment()).toMatchSnapshot();
});

test('should not set amount on invalid input', async () => {
    const { asFragment } = render(<BrowserRouter> <CreateAndEditExpense/> </BrowserRouter>);
    const amount = screen.getByPlaceholderText('Amount');
    const value = '12.777';

    
    fireEvent.change(amount, {target: {value}});
    expect(amount.value).toBe("");
    expect(asFragment()).toMatchSnapshot();
});

test('should call onsubmit prop for valid form submission', async () => {
    const submitSpy = jest.fn();

    render(<BrowserRouter> <CreateAndEditExpense 
        expenseToEdit={expenses[0]} 
        onSubmit={submitSpy}/> 
    </BrowserRouter>);
    const user = userEvent.setup();
    const submitButton = screen.getByTestId('form-submit');

    await user.click(submitButton);
    expect(submitSpy).toHaveBeenCalled();
    expect(submitSpy).toHaveBeenCalledWith({
        description: expenses[0].description,
        amount: expenses[0].amount,
        createdAt: expenses[0].createdAt.valueOf(),
        note: expenses[0].note
    })
});
