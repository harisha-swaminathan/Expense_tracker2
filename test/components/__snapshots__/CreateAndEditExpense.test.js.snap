// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expenseList is rendered with expenses as expected 1`] = `
<DocumentFragment>
   
  <form
    data-testid="form-element"
  >
    <input
      placeholder="Description"
      type="text"
      value="celine bag"
    />
    <input
      placeholder="Amount"
      type="text"
      value="8.5"
    />
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container react-datepicker__view-calendar-icon"
      >
        <svg
          class="react-datepicker__calendar-icon "
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
          />
        </svg>
        <input
          class=""
          type="text"
          value="07/19/1996"
        />
      </div>
    </div>
    <textarea
      placeholder="add a note for you expense"
    >
      bday gift to myself
    </textarea>
    <button
      data-testid="form-submit"
    >
       Add Expense
    </button>
  </form>
   
</DocumentFragment>
`;

exports[`expenseList is rendered without any expense as expected 1`] = `
<DocumentFragment>
   
  <form
    data-testid="form-element"
  >
    <input
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      placeholder="Amount"
      type="text"
      value=""
    />
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container react-datepicker__view-calendar-icon"
      >
        <svg
          class="react-datepicker__calendar-icon "
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
          />
        </svg>
        <input
          class=""
          type="text"
          value="07/19/1996"
        />
      </div>
    </div>
    <textarea
      placeholder="add a note for you expense"
    />
    <button
      data-testid="form-submit"
    >
       Add Expense
    </button>
  </form>
   
</DocumentFragment>
`;

exports[`should not set amount on invalid input 1`] = `
<DocumentFragment>
   
  <form
    data-testid="form-element"
  >
    <input
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      placeholder="Amount"
      type="text"
      value=""
    />
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container react-datepicker__view-calendar-icon"
      >
        <svg
          class="react-datepicker__calendar-icon "
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
          />
        </svg>
        <input
          class=""
          type="text"
          value="07/19/1996"
        />
      </div>
    </div>
    <textarea
      placeholder="add a note for you expense"
    />
    <button
      data-testid="form-submit"
    >
       Add Expense
    </button>
  </form>
   
</DocumentFragment>
`;

exports[`should render error 1`] = `
<DocumentFragment>
   
  <p>
    Please provide Description and amount!
  </p>
  <form
    data-testid="form-element"
  >
    <input
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      placeholder="Amount"
      type="text"
      value=""
    />
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container react-datepicker__view-calendar-icon"
      >
        <svg
          class="react-datepicker__calendar-icon "
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
          />
        </svg>
        <input
          class=""
          type="text"
          value="07/19/1996"
        />
      </div>
    </div>
    <textarea
      placeholder="add a note for you expense"
    />
    <button
      data-testid="form-submit"
    >
       Add Expense
    </button>
  </form>
   
</DocumentFragment>
`;

exports[`should set amount on valid input 1`] = `
<DocumentFragment>
   
  <form
    data-testid="form-element"
  >
    <input
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      placeholder="Amount"
      type="text"
      value="12.50"
    />
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container react-datepicker__view-calendar-icon"
      >
        <svg
          class="react-datepicker__calendar-icon "
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
          />
        </svg>
        <input
          class=""
          type="text"
          value="07/19/1996"
        />
      </div>
    </div>
    <textarea
      placeholder="add a note for you expense"
    />
    <button
      data-testid="form-submit"
    >
       Add Expense
    </button>
  </form>
   
</DocumentFragment>
`;

exports[`should set description on input change 1`] = `
<DocumentFragment>
   
  <form
    data-testid="form-element"
  >
    <input
      placeholder="Description"
      type="text"
      value="new description"
    />
    <input
      placeholder="Amount"
      type="text"
      value=""
    />
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container react-datepicker__view-calendar-icon"
      >
        <svg
          class="react-datepicker__calendar-icon "
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
          />
        </svg>
        <input
          class=""
          type="text"
          value="07/19/1996"
        />
      </div>
    </div>
    <textarea
      placeholder="add a note for you expense"
    />
    <button
      data-testid="form-submit"
    >
       Add Expense
    </button>
  </form>
   
</DocumentFragment>
`;

exports[`should set notes on input change 1`] = `
<DocumentFragment>
   
  <form
    data-testid="form-element"
  >
    <input
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      placeholder="Amount"
      type="text"
      value=""
    />
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container react-datepicker__view-calendar-icon"
      >
        <svg
          class="react-datepicker__calendar-icon "
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
          />
        </svg>
        <input
          class=""
          type="text"
          value="07/19/1996"
        />
      </div>
    </div>
    <textarea
      placeholder="add a note for you expense"
    >
      a new bag for my bday
    </textarea>
    <button
      data-testid="form-submit"
    >
       Add Expense
    </button>
  </form>
   
</DocumentFragment>
`;
